src/conduit-lwt-unix/conduit_lwt_server.ml

```diff
@@ -71,34 +71,41 @@ let run_handler handler v =
                   f "Uncaught exception in handler: %s" (Printexc.to_string ex)));
           Lwt.return_unit))
 
-let init ?(stop = fst (Lwt.wait ())) handler fd =
+let init ?(nconn = 10_000) ?(stop = fst (Lwt.wait ())) handler fd =
   let stop = Lwt.map (fun () -> `Stop) stop in
+  let log_exn = function
+    | Some ex ->
+        Log.warn (fun f ->
+            f "Uncaught exception accepting connection: %s"
+              (Printexc.to_string ex))
+    | None -> ()
+  in
   let rec loop () =
-    Lwt.try_bind
+    let accepted =
+      Lwt_unix.accept_n fd nconn >>= fun (connections, maybe_error) ->
+      log_exn maybe_error;
+      Lwt.return (`Accepted connections)
+    in
+    Lwt.catch
       (fun () ->
-        connected ();
-        throttle () >>= fun () ->
-        let accept = Lwt.map (fun v -> `Accept v) (Lwt_unix.accept fd) in
-        Lwt.choose [ accept; stop ] >|= function
-        | `Stop ->
-            Lwt.cancel accept;
-            `Stop
-        | `Accept _ as x -> x)
-      (function
+        Lwt.choose [ accepted; stop ] >>= function
         | `Stop ->
-            disconnected ();
+            Lwt.cancel accepted;
             Lwt.return_unit
-        | `Accept v ->
-            run_handler handler v;
-            loop ())
-      (fun exn ->
-        disconnected ();
-        match exn with
+        | `Accepted connections ->
+            Lwt_list.iter_p
+              (fun v ->
+                connected ();
```

Comment:
There may also be a race to update connected here since I am calling `iter_p`, right?

Reply:
I don't quite see the race you're referring to here -- where is the `iter_p`?

Reply:
If I am not mistaken `iter_p` runs the iteration in parallel, and `connected ()` increments the active connection count without any protection (`let connected () = incr active`). My fear was that multiple connections executed in parallel could try to increment the variable simultaneously (and I don't want to put a mutex there).